### PR TITLE
Separate local embedding service

### PR DESCRIPTION
**Pull request by abhishekgml1991**
## Description

[Provide a brief description of the changes in this PR]

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (Is there an addition/change in API Request, Response? If yes, documentation is required)
- [ ] Performance improvement
- [ ] Code refactoring
- [ ] Security fix

## Related Issues

- Fixes #[issue number]
- Closes #[issue number]
- Related to #[issue number]

## How Has This Been Tested?

[Describe the tests you ran to verify your changes. Provide instructions so we can reproduce.]

### Test Configuration

- [ ] Unit tests pass
- [ ] Integration tests pass
- [ ] Manual testing completed
- [ ] Cross-browser testing (if applicable)
- [ ] Mobile responsiveness tested (if applicable)

## Core Functionality Testing

Please confirm that the following core functionalities are working as expected:

- [ ] **Search capabilities** are working correctly
- [ ] **Knowledge search** is functioning properly
- [ ] **Connector indexing** is working as expected
- [ ] **Citations** are displaying and linking correctly
- [ ] **Documentation** is updated at https://docs.pipeshub.com/introduction

## What You Have Tested

Please describe what you have specifically tested:

- [ ] Feature works in development environment
- [ ] Feature works in staging environment
- [ ] Error handling scenarios tested
- [ ] Edge cases considered and tested
- [ ] Performance impact assessed
- [ ] Security implications reviewed

### Test Results

[Provide details of your test results here]

## Screenshots/Videos

**Before:**
[Attach screenshots/videos showing the current behavior]

**After:**
[Attach screenshots/videos showing the new behavior]

**Note:** Screenshots are required before merge for UI changes.

## Code Quality Checklist

- [ ] My code follows the project's coding standards
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Documentation

- [ ] Documentation has been updated (if applicable)
- [ ] API documentation updated (if applicable)
- [ ] README updated (if applicable)
- [ ] Changelog updated (if applicable)

## Security Considerations

- [ ] No sensitive data is exposed
- [ ] Input validation is implemented where necessary
- [ ] Authentication/authorization is properly handled
- [ ] No security vulnerabilities introduced

## Breaking Changes

- [ ] This PR introduces breaking changes
- [ ] Migration guide provided (if applicable)
- [ ] Version bump required

If breaking changes are introduced, please describe them here:

[Describe any breaking changes and how users should adapt]

## Performance Impact

- [ ] No performance impact
- [ ] Performance improved
- [ ] Performance impact assessed and acceptable

[Describe any performance implications]

## Dependencies

- [ ] No new dependencies added
- [ ] New dependencies are necessary and approved
- [ ] Dependencies updated and tested

List any new dependencies:
- [Dependency name and version]

## Deployment Notes

[Any special deployment considerations or steps]

## Checklist Before Merge

- [ ] All tests are passing
- [ ] Code review completed and approved
- [ ] Documentation updated and reviewed
- [ ] Screenshots/videos attached for UI changes
- [ ] Core functionality verified
- [ ] Security review completed (if applicable)
- [ ] Performance impact assessed
- [ ] Ready for production deployment

## Additional Notes

[Any additional information that would be helpful for reviewers]

---

**For Reviewers:**

Please ensure all checklist items are completed before approving this PR. Pay special attention to:
- Core functionality testing
- Security implications
- Performance impact
- Documentation completeness

### CODE REVIEW

File: backend/python/app/embedding_main.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+from app.config.constants.ai_models import (
+    DEFAULT_EMBEDDING_MODEL,
+    EMBEDDING_SERVER_PORT,
+)
+from app.utils.logger import create_logger
+from app.utils.time_conversion import get_epoch_timestamp_in_ms
+
+logger = create_logger("embedding_service")
+
+DEFAULT_DEVICE = os.getenv("EMBEDDING_SERVER_DEVICE", "cpu")
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/python/app/embedding_main.py:32]: only CPU is supported (dependencies installed)

File: backend/python/app/embedding_main.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                trust_remote_code,
+            )
+
+            def _load() -> Any:
+                from sentence_transformers import SentenceTransformer
+
+                load_kwargs: dict[str, Any] = {"device": self._device}
+                if trust_remote_code:
+                    load_kwargs["trust_remote_code"] = True
+                return SentenceTransformer(model_name, **load_kwargs)
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/python/app/embedding_main.py]: only cpu mode supported

File: backend/python/app/utils/embedding_server_client.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+async def _await_with_retry(
+    fn: Callable[[], Awaitable[T]],
+    *,
+    max_retries: int,
+    operation: str,
+) -> T:
+    last_exc: BaseException | None = None
+    for attempt in range(1, max_retries + 1):
+        try:
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/python/app/utils/embedding_server_client.py:126]: fixed

File: backend/python/app/utils/embedding_server_client.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+def _call_with_retry(
+    fn: Callable[[], T],
+    *,
+    max_retries: int,
+    operation: str,
+) -> T:
+    last_exc: BaseException | None = None
+    for attempt in range(1, max_retries + 1):
+        try:
+            return fn()
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/python/app/utils/embedding_server_client.py]: fixed

File: backend/python/pyproject.toml [hunk 5]
```diff
@@ -232,6 +233,7 @@ dependencies = [
     "lxml>=4.9.0",
     "mail-parser-reply==1.36",
     "markdown==3.8.1",
+    "megablocks>=0.10.0",
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/python/pyproject.toml]: removed

File: backend/nodejs/apps/src/modules/tokens_manager/routes/health.routes.ts [hunk 6]
```diff
@@ -276,20 +276,23 @@ export function createHealthRouter(
...
+  // Combined services health check (Python query + connector + indexing + docling + embedding services)
   router.get('/services', async (_req, res, _next) => {
     try {
       const aiHealthUrl = `${appConfig.aiBackend}/health`;
       const connectorHealthUrl = `${appConfig.connectorBackend}/health`;
       const indexingHealthUrl = `${appConfig.indexingBackend}/health`;
       const doclingBackend = process.env.DOCLING_BACKEND || 'http://localhost:8081';
       const doclingHealthUrl = `${doclingBackend}/health`;
+      const embeddingBackend = process.env.EMBEDDING_SERVER_URL || 'http://localhost:8002';
+      const embeddingHealthUrl = `${embeddingBackend}/health`;
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/nodejs/apps/src/modules/tokens_manager/routes/health.routes.ts]: Fixed

File: backend/python/app/utils/embedding_server_client.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        model: str | None = None,
+        max_retries: int | None = None,
+        timeout: float | None = None,
+        trust_remote_code: bool = False,
+    ) -> None:
+        self.model = model or DEFAULT_EMBEDDING_MODEL
+        self.max_retries = max_retries if max_retries is not None else _embedding_server_max_retries()
+        self.timeout = timeout if timeout is not None else _embedding_server_timeout()
+        self.trust_remote_code = trust_remote_code
+        extra_body = {"trust_remote_code": True} if trust_remote_code else None
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/python/app/utils/embedding_server_client.py:162]: ignore

File: backend/python/app/embedding_main.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+            trust_remote_code=request.trust_remote_code,
+        )
+    except Exception as exc:
+        logger.exception("Embedding failed for model=%s", request.model)
+        raise HTTPException(
+            status_code=500,
+            detail=f"Failed to generate embeddings: {exc}",
+        ) from exc
+
+    token_estimate = sum(max(1, len(t.split()) * 4 // 3) for t in texts)
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/python/app/embedding_main.py:348]: wrong

File: backend/python/app/embedding_main.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+
+def _format_embedding_vector(
+    vector: list[float],
+    encoding_format: str,
+) -> list[float] | str:
+    """Format a single embedding per OpenAI encoding_format semantics."""
+    if encoding_format == "base64":
+        packed = struct.pack(f"{len(vector)}f", *vector)
+        return base64.b64encode(packed).decode("ascii")
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/python/app/embedding_main.py:234]: fixed

File: deployment/docker-compose/docker-compose.build.neo4j.yml [hunk 10]
```diff
@@ -156,21 +157,24 @@ services:
...
     # is considered ready. This makes `docker compose up --wait` block until
-    # all 5 pipeshub services inside the container have finished starting.
+    # all pipeshub services inside the container have finished starting.
     healthcheck:
       test:
         - CMD-SHELL
         - >
           curl -sf http://localhost:3000/api/v1/health/services -o /tmp/hc.json &&
-          python3 -c "import json,sys; d=json.load(open('/tmp/hc.json')); s=d.get('services',{}) or {}; required=('query','connector','indexing','docling'); sys.exit(0 if all(s.get(k)=='healthy' for k in required) else 1)"
+          python3 -c "import json,sys; d=json.load(open('/tmp/hc.json')); s=d.get('services',{}) or {}; required=('query','connector','indexing','docling','embedding'); sys.exit(0 if all(s.get(k)=='healthy' for k in required) else 1)"
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply deployment/docker-compose/docker-compose.build.neo4j.yml]: Fixed

File: deployment/docker-compose/docker-compose.build.arango.yml [hunk 11]
```diff
@@ -160,21 +161,24 @@ services:
...
     # is considered ready. This makes `docker compose up --wait` block until
-    # all 5 pipeshub services inside the container have finished starting.
+    # all pipeshub services inside the container have finished starting.
     healthcheck:
       test:
         - CMD-SHELL
         - >
           curl -sf http://localhost:3000/api/v1/health/services -o /tmp/hc.json &&
-          python3 -c "import json,sys; d=json.load(open('/tmp/hc.json')); s=d.get('services',{}) or {}; required=('query','connector','indexing','docling'); sys.exit(0 if all(s.get(k)=='healthy' for k in required) else 1)"
+          python3 -c "import json,sys; d=json.load(open('/tmp/hc.json')); s=d.get('services',{}) or {}; required=('query','connector','indexing','docling','embedding'); sys.exit(0 if all(s.get(k)=='healthy' for k in required) else 1)"
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply deployment/docker-compose/docker-compose.build.arango.yml]: Fixed

File: backend/python/tests/unit/test_embedding_main.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+                "model": DEFAULT_EMBEDDING_MODEL,
+                "input": "hello",
+                "encoding_format": "base64",
+            },
+        )
+        assert response.status_code == 200
+        body = response.json()
+        expected = base64.b64encode(
+            struct.pack("3f", 0.1, 0.2, 0.3)
+        ).decode("ascii")
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/python/tests/unit/test_embedding_main.py:98]: Fixed

File: backend/python/app/embedding_main.py [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+# library is imported inside ``_load``.
+os.environ.setdefault("HF_HUB_DOWNLOAD_TIMEOUT", "30")
+
+
+class EmbeddingRequest(BaseModel):
+    model: str
+    input: str | list[str]
+    encoding_format: str = "float"
+    user: str | None = None
+    trust_remote_code: bool = False
```
[the remark this answers is not in the document]
Comment (abhishek-pipeshub) [reply backend/python/app/embedding_main.py:74]: Fixed

**Review by abhishek-pipeshub — APPROVED** [review]
